File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> result = new ArrayList<>();

        int min_row = 0;
        int min_col = 0;
        int max_row = matrix.length-1;
        int max_col = matrix[0].length-1;

        while(min_row <= max_row && min_col <= max_col){
            for(int j = min_col ; j <= max_col ; j++  ){
                result.add(matrix[min_row][j]);
            }
            
            for(int i = min_row+1 ; i <= max_row ; i++){
                result.add(matrix[i][max_col]);
            }
            
            if(min_row < max_row){ 
            for(int j = max_col-1 ; j >= min_col ; j--){
                result.add(matrix[max_row][j]);
            }
            }
            
            if(min_col < max_col){ 
            for(int i = max_row-1 ; i >= min_row+1; i--){
                result.add(matrix[i][min_col]);
            }
            }
            min_col++;
            min_row++;
            max_col--;
            max_row--;
        }
        return result;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ans ;

        int start_row = 0;
        int start_col = 0;
        int end_row = matrix.size()-1;
        int end_col = matrix[0].size()-1;

        while(start_row <= end_row && start_col <= end_col){
            for(int j = start_col ; j <= end_col; j++){
                ans.push_back(matrix[start_row][j]);
            }

            for(int i = start_row+1; i <= end_row ; i++){
                ans.push_back(matrix[i][end_col]);
            }

            if(start_row != end_row){
                for(int j = end_col -1 ; j >= start_col ; j--){
                    ans.push_back(matrix[end_row][j]);
                }
            }
            if(start_col != end_col){
                for(int i = end_row-1 ; i >= start_row+1; i--){
                    ans.push_back(matrix[i][start_col]);
                }


            }

            start_row++;
            end_row--;
            end_col--;
            start_col++;
        }

        return ans;
        
        
    }
};